extensions/amp-bind/0.1/bind-evaluator.js

```diff
@@ -121,15 +122,27 @@ export class BindEvaluator {
         errors[expressionString] = {message: error.message, stack: error.stack};
         return;
       }
+      cache[expressionString] = result;
+    });
+
+    // Then, validate each binding and delete invalid expression results.
+    this.bindings_.forEach(binding => {
+      const {tagName, property, expressionString} = binding;
+      const result = cache[expressionString];
+      if (result === undefined) {
+        return;
+      }
       const resultString = this.stringValueOf_(property, result);
-      if (this.validator_.isResultValid(tagName, property, resultString)) {
-        cache[expressionString] = result;
-      } else {
+      if (!this.validator_.isResultValid(tagName, property, resultString)) {
```

Comment:
I think the `delete` is necessary in either case and will only be triggered on invalid results (should be rare).

IMO the readability trade off is worth it, especially in security-sensitive contexts. Here's the single loop version for comparison:
```
this.bindings_.forEach(binding => {
  const {tagName, property, expressionString} = binding;

  let result = cache[expressionString];
  let error = errors[expressionString];

  if (result === undefined && !error) {
    const expression = this.expressions_[expressionString];
    if (expression) {
      const evaluation = this.evaluate_(expression, scope);
      if (evaluation.error) {
        error = {
          message: evaluation.error.message, 
          stack: evalulation.error.stack,
        };
      } else {
        result = evaluation.result;
      }
    } else {
      const e =
          new Error(`Expression "${expressionString}"" is not cached.`);
      error = {message: e.message, stack: e.stack};
    }
    errors[expressionString] = error;
  }

  if (result !== undefined && !error) {
    const resultString = this.stringValueOf_(property, result);
    if (this.validator_.isResultValid(tagName, property, resultString)) {
      cache[expressionString] = result;
    } else {
      // TODO(choumx): If this expression string is used in another
      // tagName/property which is valid, we ought to allow it.
      delete cache[expressionString];
      const e =
          new Error(`"${result}" is not a valid result for [${property}].`);
      errors[expressionString] = {message: e.message, stack: e.stack};
    }
  }
});
```